File: datastructure (& related algorithm)/queue.py
# Definition:
# - A queue of elements of type T is a finite sequence of elements of T,
# in which data can only be inserted at one end called the rear, and
# deleted from the other end called the front.
# - Queue is a First In - First Out (FIFO) data structure.
# (FIFO: The first item stored in the queue is the first item that can be
# taken out.)

MAX_SIZE_QUEUE = 1000

class Node:
    def __init__(self, data, next = None):
        self.data = data
        self.next = next

class Queue:
    def __init__(self, head = None, tail = None):
        self.__head = head
        self.__tail = tail
        self.__count = 0

    def dequeue(self):
        if self.isEmpty():
            return None
        head = self.__head
        self.__head = head.next
        self.__count -= 1
        return head.data
    
    def enqueue(self, data):
        if self.isFull():
            return False
        tail = Node(data)
        if self.isEmpty():
            self.__head = self.__tail = tail
        else:
            self.__tail.next = tail
            self.__tail = tail
        self.__count += 1
        return True

    def isFull(self):
        return self.__count >= MAX_SIZE_QUEUE

    def getQueueFront(self):
        return self.__head

    def getQueueRear(self):
        return self.__tail

    def isEmpty(self):
        return self.__count == 0

    def getSize(self):
        return self.__count

    def printQueue(self):
        head = self.__head
        print("-"*20 + "\nSize: " + str(self.__count) + "\nQueue:",end=" ")
        while head is not None:
            print(head.data, end = " ")
            head = head.next
        print("\n" + "-"*20)

File: datastructure (& related algorithm)/queue.cpp
// Definition:
// - A queue of elements of type T is a finite sequence of elements of T,
// in which data can only be inserted at one end called the rear, and
// deleted from the other end called the front.
// - Queue is a First In - First Out (FIFO) data structure.
// (FIFO: The first item stored in the queue is the first item that can be
// taken out.)

#include "list.h"

template <class ItemType>
class Queue : public List<ItemType>{
    Node<ItemType> *tail;
public:
    Queue(Node<ItemType>* head = nullptr) : List<ItemType>(head){
        this->tail = head;
    }
    ~Queue() {}
    ItemType dequeue(){
        if(this->isEmpty()){
            cout<<"Queue is empty.\n";
            return (ItemType)NULL;
        }
        Node<ItemType>* head = this->head;
        this->head = head->getNext();
        this->count--;
        return head->getData();
    }
    bool enqueue(ItemType data){
        if(this->isFull()){
            cout<<"Queue is full.\n";
            return false;
        }
        if(this->isEmpty()){
            this->head = new Node<ItemType>(data);
            this->tail = this->head;
        }
        else{
            Node<ItemType> *tail = new Node<ItemType>(data);
            this->tail->setNext(tail);
            this->tail = tail;
        }
        this->count++;
        return true;

    }
    Node<ItemType> *getQueueFront(){
        return List<ItemType>::getHead();
    }
    Node<ItemType> * getQueueRear(){
        return this->tail;
    }
};

// int main(int argc, char const *argv[])
// {
//     Queue<int> * queue = new Queue<int>();
//     queue->enqueue(7);
//     queue->enqueue(8);
//     queue->enqueue(12);
//     queue->print();
//     queue->dequeue();
//     queue->print();
//     return 0;
// }

